### DISCUSSION TITLE
this.gui.setContent(); alternative and problem with addItems

### DISCUSSION THREAD

**Discussion by Challangerson** [NichtStudioCode/InvUI#59] [Q&A] [answered]
![image](https://github.com/NichtStudioCode/InvUI/assets/124780390/d94ea315-cfe6-480c-99d6-2425c3756f94)

if(!controlItems.isEmpty()) {
            this.gui.setContent(controlItems);
}

    public BansGui(LogsManager logsManager) {
        this.logsManager = logsManager;
        this.gui = PagedGui.items()
                .setStructure(
                        "# # # # # # # # #",
                        "# x x x x x x x #",
                        "# x x x x x x x #",
                        "# x x x x x x x #",
                        "# # # < # > # # #"
                ).addIngredient(
                        '#', new SimpleItem(new ItemBuilder(Material.BLACK_STAINED_GLASS_PANE).setDisplayName(" "))
                ).addIngredient(
                        '<', new BackItem()
                ).addIngredient(
                        'x', Markers.CONTENT_LIST_SLOT_HORIZONTAL
                ).addIngredient(
                        '>', new ForwardItem()
                ).build();
    }

    private void prepareGui() {
        for(Logs logs : this.logsManager.getLogs().values()) {

            this.gui.addItems(
                new ControlItem<>() {

                    @Override
                    public void handleClick(@NotNull ClickType clickType, @NotNull Player player, @NotNull InventoryClickEvent event) {
                        if(clickType.isLeftClick()) {
                            ChatUtil.sendMessage(player, "&8>> &cPowod: &7" + logs.getReason());
                        }
                    }

                    @Override
                    public ItemProvider getItemProvider(Gui gui) {

                        return new ItemBuilder(logs.getLogsType().getMaterial())
                        .setDisplayName(ChatUtil.fixColor("&f>> &e" + logs.getPunishment() + " &f<<"))
                        .setLegacyLore(
                            ChatUtil.fixColors(
                                Arrays.asList(
                                        "&8>> &7Rodzaj kary: &c" + logs.getLogsType().getName(),
                                        "&8>> &7Administrator: &c" + logs.getPunisher(),
                                        "&8>> &7Powod: &c" + logs.getReason(),
                                        "&8>> &7Czas rozpoczecia: &c" + TimeUtil.getDateFromLong(logs.getTimeStarted().toEpochMilli()),
                                        "&8>> &7Czas: &c" + (logs.getTime() >= 0
                                                ? "PERMANENTNIE" : (logs.getTime() > System.currentTimeMillis()
                                                ? TimeUtil.secondsToString(logs.getTime() - System.currentTimeMillis()) : "zakonczony")
                                )
                            )
                        ));
                    }

                }
            );
            

        }
    }

    public void openGui(Player player) {
        Window window = Window.single()
                .setTitle(ChatUtil.fixColor("&f>> &4Historia banow &f<<"))
                .setViewer(player)
                .setGui(gui)
                .build();

        this.prepareGui();
        window.open();
    }
    
    
how can i replace a setItems for no dupling

**Comment by NichtStudioCode** [comment]
What exactly is your question?

**Reply by Challangerson** [reply]
My question is, how can I add items to 'inv' without duplicating them? The 'addItems' function doesn't work exactly as I want; it adds items every time I open the GUI. I want to add items only once every time someone opens the inventory. The logs have two elements, and when a command adds one to the logs, the logs have three elements. In the GUI, it displays three items every time if the next ones are not added.

**Reply by NichtStudioCode** [reply]
Just only call Gui#addItems once?
You're also probably looking for PagedGui#setContent.
I'm also not sure why you're using ControlItems here.

Please make sure to read the documentation: https://docs.xen.cx/invui

**Reply by Challangerson** [reply]
![image](https://github.com/NichtStudioCode/InvUI/assets/124780390/38ce15a3-da66-4c57-86c7-68ed30a95f6c)

**Reply by NichtStudioCode** [reply, marked as answer]
Don't use raw types
